Publish current metrics for all services to statsd adapter using the new Memory Mapped File structure

mgherghe opened the pull request.

Commit:
Refactored metrics.reader to implement the new MMF structure;
Refactored metrics.viewer and metrics.statsD to use the new reader implementation;
Minor changes to whitespaces in some files

Commit:
Minor pom changes to fix warning messages

Commit:
Added k3po dependency so k3po (robot) can be started from reader subdirectory using command:
mvn -Dmaven.robot.daemon=false robot:start
Added a skeleton new k3po test and script and utility to create an MMF from script output.

Commit:
Changed FileWriter into a rule, WriteToFileRule, to make it easier to write tests.
Filled out the robot test, MonitoringDataProcessorIT,a bit. It's now close to working. Of course more test cases need to be added.
Made a couple of changes to thre reader classes to get the test working:
- in AgronaMonitoringDataProcessor.getMappedFile(), allow use of an absolute fileName
- in MMFReaderBuilderImpl.readMetadata(int), don't use METADATA_BUFFER_LENGTH because the file may be smaller

Commit:
Added unit tests;
Fixed minor bug in GatewayCountersImpl;
Changed the format of the name of the counters sent to StatsD

Commit:
Merge pull request #1 from cmebarrow/add_k3po_to_marina_work

Add k3po to marina's work

Commit:
Added conditions to metrics.reader to cover the case where there are no gateway / service counters;
Minor formatting changes;
Added ITest for a MMF with services (not ready yet)

Commit:
Changed Agrona version from 0.3.1 to 0.4.2
Fixed ITest to generate and test a MMF with service counters
Removed unused variable from MMFReaderBuilderImpl

cmebarrow commented on reader/src/main/java/org/kaazing/monitoring/reader/api/GatewayCounters.java:
```diff
@@ -0,0 +1,41 @@
+/**
```
Doesn't look as if this class is used anywhere. If it's dead code please remove it.

mgherghe replied:
This is used in the viewer and statsD publisher. (for the moment, the gateway does not expose any gateway counters, but it will, in the future)

cmebarrow replied:
I can't find any usage in  the viewer and statsD publisher. But anyway I think your idea is to have GatewayCounters as a logical peer of ServiceCounters? I think that is good, and in that case you should replace MMFReader methods getGatewayID and getGatewyCounters with a single method GatewayCounters.getGateway().

I like the idea of having the org.kaazing.monitoring.reader.api package. It would be good to have that be the only external API for use by other projects (viewer and statsd), and have it contain only classes or interfaces necessary for that. I have submitted pull request https://github.com/mgherghe/metrics/pull/2 to your repository with these changes.

cmebarrow commented on reader/src/main/java/org/kaazing/monitoring/reader/impl/MetricsCollectorAgrona.java:
```diff
@@ -41,19 +41,20 @@ public MetricsCollectorAgrona(CountersManagerEx countersManager) {
-    public List<Metric> getMetrics() {
-        List<Metric> metrics = new ArrayList<Metric>();
+    public List<Counter> getCounters() {
+        List<Counter> counters = new ArrayList<Counter>();
```
@mgherghe Please address this comment (my PR does not address it):We can't afford to have this method create objects because since the method is called repeatedly it will cause a lot of garbage collection and reduce performance. The aim should be that each iteration  of viewer or statsd creates no objects at all . Here's a proposal to avoid any object creation from this method:
- create the list of counters in the constructor and store it in a member variable (counters)
- getCounters then just returns counters
- change CounterImpl so it can dynamically read the latest value from valuesBuffer. That can be done by changing the constructor to the following:

```
  CounterImpl(String label, int id, CountersManagerEx countersManager)
```

 and having the getValue method do 

```
return countersManager.getLongValueForId(id);
```

cmebarrow commented:
I have a general comment about the design of the reader API. Other projects should not need to reference classes in the impl package (like AgronaMonitoringDataProcessor). And please try to simplify the API. I would suggest eliminating the MetricsCollector and MonitoringDataProcessor interfaces so users need only call MMFReaderBuilder, which can be made into a concrete class taking the file name as a constructor parameter. Usage would be something like this (from statsd MainApp):
```
MMFReaderBuilder builder = new MMFReaderBuilder(args[0]);
MMFReader reader;
        try {
            while ((reader = builder.build()) == null) {
                    Thread.sleep(updateInterval);
        } catch (MetricsReaderException e) {
            LOGGER.error("There was a problem with the metrics.reader configuration file. Exiting application.");
            System.exit(1);
        }
```

cmebarrow commented on reader/src/main/java/org/kaazing/monitoring/reader/impl/ServiceCountersImpl.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public List<Counter> getCounters() {
+        MetricsCollector metricsCollector = new MetricsCollectorAgrona(countersManager);
```
@mgherghe Please address this comment (my PR does not address it):To avoid object creation and garbage collection the MetricsCollectorAgrona object should be created just once in the constructor.

cmebarrow commented on reader/src/main/java/org/kaazing/monitoring/reader/api/MMFReader.java:
```diff
@@ -19,37 +19,44 @@
+     * @param service
+     * @return List<Counter>
+     */
+    List<Counter> getServiceCounters(ServiceCounters service);
```
This method seems to overlap a bit with getServices(). And the name implies it should return a ServiceCounters instance. I would suggest either removing this method or changing it to take the service name as the input parameter and return ServiceCounters:

```
ServiceCounter getServiceCounters(String serviceName)
```

mgherghe replied:
This method is used to retrieve all the counters for a single service. I have renamed it to "getCountersFromService" to be more suggestive. I don't think a getServiceCounters method that would return a ServiceCounters instance could be used anywhere.

cmebarrow replied:
That makes more sense but I don't think this method is useful. All it does is call service.getCounters() which the caller of the method can easily do themselves. So this method is redundant. Please remove it and replace all usages with service.getCounters(). I have made that change in the pull request I submitted to you.

cmebarrow commented:
@mgherghe Hi Marina, I have not managed to look in detail at all of the code yet (it's getting too late!) but I've given enough feedback to keep you busy for a bit. I will try to look at the rest of it tomorrow. Most of it is looking good!

cmebarrow commented on viewer/src/main/java/org/kaazing/monitoring/agrona/viewer/task/MetricsTaskImpl.java:
```diff
@@ -13,16 +14,20 @@
+            // metrics retrieved - log message directly added to the logger output for each metric (using default
+            // logging in metricsCollector.getServiceCounters() method)
+                for (ServiceCounters service : reader.getServices()) {
+                    reader.getServiceCounters(service);
```
Just call service.getCounters() instead of reader.getServiceCounters(service); then you can get rid of the getServiceCounters method.

Commit:
- Move MMFReaderBuilder, MetricsCollector, and MonitoringDataProcessor from reader api package to impl.
- Add static wrap method to MMFReader interface.
- Replace MMFReader methods getGatewayID and getGatewayCounters with getGateway (returning GatewayCounters).
- Rename GatewayCounters method getGatewayId to getId.
- Change AgronaMonitoringDataProcessor to use MMFReader.wrap().

Commit:
- Rename MMFReader to Metrics and tweak the javadoc
- Move MonitoringDataProcessor back to api package, rename to MetricsFileProcessor
  and add newInstance method
- Rename MonitoringFolderAgrona to MonitoringFolder
- Rename MonitoringFolderAgronaImpl to MonitoringFolderImpl
- Rename related unit test classes

cmebarrow commented:
@mgherghe Hi Marina, I found I was making a lot of cosmetic comments about the API (mainly just to rename some things) so thought it would be easier to subject a pull request: https://github.com/mgherghe/metrics/pull/2. You can just merge it to your repository if you are okay with the proposed changes (and you can then always tweak anything you don't like). This PR addresses all my cosmetic comments.

@mgherghe Please note that I have not addressed the two comments I made about garbage collection (on classes MetricsCollectorAgrona and ServiceCountersImpl, so please work on that). Once that's done I'll be able to merge your work.

Commit:
Moved creating object lists from methods to the constructor (in MetricsCollectorAgrona, ServiceCountersImpl and GatewayCountersImpl)

Commit:
Rename GatewayCounters.getId back to getGatewayId.

Commit:
Merge remote-tracking branch 'marina/develop' into metrics_api_polishing

Commit:
Fix compile error in MetricsCollectorAgrona (wrong import). Remove bin directories which were accidentally added.

Commit:
Change reader/pom.xml to add failsave plugin so the robot tests actually get executed during build! Fix an issue in one robot test and add TODO comment to it.

Commit:
Merge pull request #2 from cmebarrow/metrics_api_polishing

Metrics api polishing

Commit:
Minor formatting and refactoring
Moved internal interfaces from impl to interfaces package
Added debugging message to metrics.viewer when there is no file
Modified ITest to also check the counters' values

cmebarrow commented:
@mgherghe Looking good, I am just waiting for Irina to complete then I will merge (I want to merge them together so things stay consistent).

Commit:
Update Readme file for metrics.statsd
